### PR TITLE
Aql parser

### PR DESCRIPTION
**Pull request by JavierJia**
Implement a preliminary query parser. Most of the implementations are still not implemented. Just tried a simple query to make a POC test and verify those interfaces.

#30 
@MoniMoledo @aishwaryakapse

### CODE REVIEW

File: zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/datastore/IQueryParser.scala [hunk 1]
```diff
@@ -3,5 +3,16 @@ package edu.uci.ics.cloudberry.zion.model.datastore
...
+  /**
+    * Parser the Query to string statements. One query may produce multiple statements.
+    *
+    * @param query
+    * @param schema
+    * @return
+    */
+  def parse(query: Query, schema: Schema): Seq[String]
+
+  protected def requireOrThrow(boolean: Boolean, msg: String): Unit = {
```
Comment (JavierJia) [inline zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/datastore/IQueryParser.scala]: `boolean` to `condition`

File: zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/datastore/IResponseHandler.scala [hunk 2]
```diff
@@ -0,0 +1,7 @@
+package edu.uci.ics.cloudberry.zion.model.datastore
+
+import play.api.libs.ws.WSResponse
+
+trait IResponseHandler {
```
Comment (JavierJia) [inline zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/datastore/IResponseHandler.scala]: `IWSResponseHandler`

File: zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/AQLFuncVisitor.scala [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+    implicitly[TypeTag[T]].tpe <:< typeOf[AnyVal]
+  }
+
+  def translateRelation(field: Field,
+                        funcOpt: Option[TransformFunc],
+                        sourceVar: String,
+                        relation: Relation,
+                        values: Seq[Any]
+                       ): String = {
+    //TODO add the functions
```
Comment (JavierJia) [inline zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/AQLFuncVisitor.scala]: //TODO add the function handling logic

File: zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/AQLFuncVisitor.scala [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+    if (!Schema.Type2Relations.get(field.dataType).get.contains(relation)) {
+      throw QueryParsingException(s"field ${field.name} of type ${field.dataType} can not apply to relation: ${relation}")
```
Comment (JavierJia) [inline zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/AQLFuncVisitor.scala]: too long

File: zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/AQLFuncVisitor.scala [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+          case ex: IllegalArgumentException => throw QueryParsingException("invalid time format")
+        }
+        translateTimeRelation(field, funcOpt, sourceVar, relation, values.map(_.asInstanceOf[String]))
+      case DataType.Point =>
+        //TODO support circle and polygon
+        if (!values.forall(_.isInstanceOf[Seq[Number]]) || values.size != 2
+          || values.map(_.asInstanceOf[Seq[Number]]).forall(_.size != 2)) {
+          throw QueryParsingException(s"the ${relation} on point type requires a pair of value pairs")
+        }
+        translatePointRelation(field, funcOpt, sourceVar, relation, values.map(_.asInstanceOf[Seq[Number]].map(_.doubleValue())))
```
Comment (JavierJia) [inline zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/AQLFuncVisitor.scala]: too long

File: zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/AQLFuncVisitor.scala [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+
+  private def translateTimeRelation(field: Field,
+                                    funcOpt: Option[TransformFunc],
+                                    sourceVar: String,
+                                    relation: Relation,
+                                    values: Seq[String]): String = {
+    relation match {
+      case Relation.inRange => {
+        if (values.size != 2) throw QueryParsingException(s"relation: ${relation} require two parameters")
+        return s"$sourceVar.${field.name} >= datetime('${values(0)}') and $sourceVar.${field.name} < datetime('${values(1)}')"
```
Comment (JavierJia) [inline zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/AQLFuncVisitor.scala]: too long

File: zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/AQLFuncVisitor.scala [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+      case _ => throw QueryParsingException(s"point type doesn't support relation $relation")
+    }
+  }
+
+  private def translateTextRelation(field: Field,
+                                    funcOpt: Option[TransformFunc],
+                                    sourceVar: String,
+                                    relation: Relation,
+                                    values: Seq[String]): String = {
+    val first =s"""similarity-jaccard(word-tokens($sourceVar.'${field.name}'), word-tokens("${values.head}")) > 0.0"""
```
Comment (JavierJia) [inline zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/AQLFuncVisitor.scala]: too long

File: zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/AQLFuncVisitor.scala [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+                        ): String = {
+    funcOpt.map { func =>
+      IFunction.verifyField(func, field).map { msg => throw QueryParsingException(msg) }
+      func match {
+        case bin: Bin => ???
+        case interval: Interval =>
+          import TimeUnit._
+          //PnYnMnDTnHnMn.mmmS
+          val duration = interval.unit match {
+            case Second => s""" day-time-duration("PT${interval.x}S") """
```
Comment (JavierJia) [inline zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/AQLFuncVisitor.scala]: unify the function?

File: zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/datastore/IQueryParser.scala (hunk 1 above)
Comment (JavierJia) [inline zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/datastore/IQueryParser.scala]: done

File: zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/datastore/IResponseHandler.scala (hunk 2 above)
Comment (JavierJia) [inline zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/datastore/IResponseHandler.scala]: this is an unused file, removed.

File: zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/AQLFuncVisitor.scala (hunk 3 above)
Comment (JavierJia) [inline zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/AQLFuncVisitor.scala]: done

File: zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/AQLFuncVisitor.scala (hunk 4 above)
Comment (JavierJia) [inline zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/AQLFuncVisitor.scala]: hard to break the lines. keep it.

File: zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/AQLFuncVisitor.scala (hunk 5 above)
Comment (JavierJia) [inline zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/AQLFuncVisitor.scala]: done

File: zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/AQLFuncVisitor.scala (hunk 6 above)
Comment (JavierJia) [inline zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/AQLFuncVisitor.scala]: done

File: zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/AQLFuncVisitor.scala (hunk 7 above)
Comment (JavierJia) [inline zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/AQLFuncVisitor.scala]: done

File: zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/AQLFuncVisitor.scala (hunk 8 above)
Comment (JavierJia) [inline zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/AQLFuncVisitor.scala]: not that worthy

File: zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/AQLFuncVisitor.scala [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+            case Minute => s""" day-time-duration("PT${interval.x}M") """
+            case Hour => s""" day-time-duration("PT${interval.x}H") """
+            case Day => s""" day-time-duration("P${interval.x}D") """
+            case Week => s""" day-time-duration("P${interval.x * 7}D") """
+            case Month => s""" year-month-duration("P${interval.x}M") """
+            case Year => s""" year-month-duration("P${interval.x}Y") """
+          }
+          s"get-interval-start-datetime(interval-bin($sourceVar.${field.name}, datetime('1990-01-01T00:00:00.000Z'), $duration))"
```
Comment (JavierJia) [inline zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/AQLFuncVisitor.scala]: better not break the string statement to make it more readable.

File: zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/schema/Query.scala [hunk 10]
```diff
@@ -17,66 +17,66 @@ trait Statement {}
...
   * @param selectValues
   * @param as
   */
-class LookupStatement(val sourceKeys: Seq[String],
-                      val lookupDataset: String,
-                      val lookupKeys: Seq[String],
-                      val selectValues: Seq[String],
-                      val as: Seq[String]
-                     ) extends Statement {
+case class LookupStatement(val sourceKeys: Seq[String],
```
Comment (MoniMoledo) [inline zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/schema/Query.scala]: What does  "case" before class means?

Comment (JavierJia) [inline zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/schema/Query.scala]: `case class` is a syntax sugar for `struct data`. Here is the [doc](http://docs.scala-lang.org/tutorials/tour/case-classes.html)

Comment (JavierJia) [inline zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/schema/Query.scala]: There is one comment below that doc which explains the way we are using case class:
> I often use case classes (of course if there are no complex inheritance) to get auto-generated equals, hashCode, toString, static apply() for shorter initialization, etc.

**Pull request merged**